ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,723 @@
+          ]
+      , bgroup "LegacyDiff" [
+              bench "Benchmark interpret" $
+                nf interpret legacyCmds
+          , testProperty "Sanity check" $
+              sanityCheck model legacyCmds (interpret legacyCmds)
+          ]
+      , testProperty "Interpretations match" $
+            interpret newCmds
+          ===
+            to (interpret legacyCmds)
+      ]
+  ]
+  where
+    target :: String
+    target =
+      "    $(NF-3) == \"" <> name <> "\"      \
+      \ && $(NF-1) == \"LegacyDiff\"          \
+      \ && $NF     == \"Benchmark interpret\" "
+
+{-------------------------------------------------------------------------------
+  Commands
+-------------------------------------------------------------------------------}
+
+data Imp = Legacy | New
+
+data Cmd (i :: Imp) k v =
+    Push     (SlotNo i)      (Diff i k v)
+  | Flush    Int
+  | Rollback Int
+  | Forward  (Values i k v)  (Keys i k v)
+  deriving stock Generic
+
+-- TODO: Add trace to custom instance that shows when forced, don't keep in
+-- benchmark code.
+-- TODO: Look at output with -ddump-deriv.
+deriving anyclass instance ( NFData (DiffSeq i k v)
+                           , NFData (Diff i k v)
+                           , NFData (Values i k v)
+                           , NFData (Keys i k v)
+                           , NFData (SlotNo i)
+                           ) => NFData (Cmd i k v)
+
+deriving stock instance ( Show (DiffSeq i k v)
+                        , Show (Diff i k v)
+                        , Show (Values i k v)
+                        , Show (Keys i k v)
+                        , Show (SlotNo i)
+                        ) => Show (Cmd i k v)
+
+instance (Ord k, Eq v) => Isomorphism (Cmd New k v) (Cmd Legacy k v) where
+  to :: Cmd New k v -> Cmd Legacy k v
+  to (Push sl d)     = Push (to sl) (to d)
+  to (Flush n)       = Flush n
+  to (Rollback n)    = Rollback n
+  to (Forward vs ks) = Forward (to vs) (to ks)
+
+instance (Ord k, Eq v) => Isomorphism (Cmd Legacy k v) (Cmd New k v) where
+  to :: Cmd Legacy k v -> Cmd New k v
+  to (Push sl d)     = Push (to sl) (to d)
+  to (Flush n)       = Flush n
+  to (Rollback n)    = Rollback n
+  to (Forward vs ks) = Forward (to vs) (to ks)
+
+{-------------------------------------------------------------------------------
+  Interpreter
+-------------------------------------------------------------------------------}
+
+-- | Interpretation of @Cmd@s as a fold.
+interpret ::
+     forall i k v.
+     ( Ord k
+     , Eq v
+     , Comparable i
+     , Monoid (DiffSeq i k v)
+     , NFData (DiffSeq i k v)
+     )
+  => [Cmd i k v]
+  -> DiffSeq i k v
+interpret = foldl go mempty
+  where
+    go :: DiffSeq i k v -> Cmd i k v -> DiffSeq i k v
+    go ds = \case
+      Push sl d     -> push ds sl d
+      -- We @'deepseq'@ the left part of the flush result, because that is the
+      -- part that would be written to disk, for which it must be fully
+      -- evaluated.
```

Comment:
Time spent evaluating `l` is implementation-dependent, and since we flush we know `l` will be fully evaluated before writing `l` to disk. However, @Jasagredo pointed out that we should actually not be evaluating `l`, but rather the cumulative diff of `l`.